src/Api/Endpoints/Candidate/Visit/Visit.php

```diff
@@ -0,0 +1,285 @@
+            return new \LORIS\Http\Response\NotFound('Visit not found');
+        }
+
+        return new \LORIS\Http\Response\JsonResponse(
+            (new \LORIS\Api\Views\Visit($this->_visit))
+                ->toArray()
+        );
+    }
+
+    /**
+     * Handles a PUT request that creates or replace a candidate visit
+     *
+     * TODO: There is no way to validate the the visit_label in the url
+     *       fits the json data of the request because it is removed from the
+     *       pathparts in the calling class. The correct way to do it would be
+     *       to pass a "light" timepoint class that contains the visit_label but
+     *       no sessionid in the constructor.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePUT(ServerRequestInterface $request): ResponseInterface
+    {
+        $user      = $request->getAttribute('user');
+        $data      = json_decode((string) $request->getBody(), true);
+        $visitinfo = $data ?? array();
+
+        $requiredfields = array(
+                           'CandID',
+                           'Visit',
+                           'Site',
+                           'Battery',
+                          );
+        $diff           = array_diff($requiredfields, array_keys($visitinfo));
+        if (!empty($diff)) {
+            return new \LORIS\Http\Response\BadRequest(
+                'Field(s) missing: ' . implode(', ', $diff)
+            );
+        }
+
+        if ($visitinfo['CandID'] !== $this->_candidate->getCandID()) {
+            return new \LORIS\Http\Response\BadRequest(
+                'CandID do not match this candidate'
+            );
+        }
+
+        $sessionid = array_search(
+            $visitinfo['Visit'],
+            $this->_candidate->getListOfVisitLabels()
+        );
+
+        $centerid = array_search($visitinfo['Site'], \Utility::getSiteList());
+
+        if (!in_array($centerid, $user->getCenterIDs())) {
+            return new \LORIS\Http\Response\Forbidden(
+                'You can`t create candidates visit for that site'
+            );
+        }
+
+        $subprojectid = array_search(
+            $visitinfo['Battery'],
+            \Utility::getSubprojectList()
+        );
+
+        if ($sessionid !== false) {
+            // If the visit label already exists for that candidate
+            $timepoint = \NDB_Factory::singleton()->timepoint($sessionid);
+            if ($timepoint->getCurrentStage() !== 'Not Started') {
+                return new \LORIS\Http\Response\Conflict(
+                    'This visit is already started'
+                );
+            }
+
+            $username = $user->getUsername();
+            $today    = date("Y-m-d");
+            // TODO :: Add a replace function in \Timepoint.class.inc
+            $timepoint->setData(array('CenterID' => $centerid));
```

Comment:
I think this can all be done with one function call to `setData()`

Reply:
What can be done?

Reply:
Instead of 
```php
$timepoint->setData(array('key' => 'value'));
$timepoint->setData(array('key2' => 'value2'));
$timepoint->setData(array('key3' => 'value2'));
...
```

Do 
```php
$timepoint->setData(
    array(
        'key' => 'value',
        'key2' => 'value2',
        'key3' => 'value3',
        ...
    )
);

Reply:
Less overhead with function calls and round-trips to the DB. Might not make a difference, but who knows